File: players/dancer/world.py
"""L1 WorldModel: object tracks from the egocentric NMMO3 obs stream.

The obs entity bytes are write-only residue: a window cell's entity
bytes are rewritten every tick an entity stands on the tile under it,
and NEVER cleared. Cell-level liveness classification was measured
unreliable (v2/v3); this tracker works at the object level instead:

- A track is created/confirmed ONLY by an observed byte change (a
  change at a window offset proves a write, which proves presence).
- Between confirmations a track is propagated by the exact chase model
  (melee within aggro range chase us deterministically - verified
  33/33) or held stationary, aging until expiry.
- Imprints with no track are residue and are exposed nowhere.

Frame: cumulative-offset coordinates anchored at this life's spawn.
Our own position advances by the verified own-shift rule (anim proves
last move success + speed; the action supplies direction). Teleportitis
breaks the anchor - detected via a changed-cell explosion and handled
by a full reset.

Consumers get window-relative views each tick: melee tracks, bow
tracks (element byte != 0 <=> level >= 15 <=> ranged), player tracks,
and an occupancy set for move planning.
"""

from __future__ import annotations

from dataclasses import dataclass, field

import numpy as np

from ..scripted_player import (ANIM_DEATH, CENTER_COL, CENTER_ROW,
                               ENTITY_ENEMY, ENTITY_PLAYER, MOVE_DELTAS,
                               Percept, RUN_OFFSET, WINDOW_COLS, WINDOW_ROWS)

ANIM_IDLE = 0
ANIM_MOVE = 1
ANIM_RUN = 6

NPC_AGGRO = 4
TELEPORT_CHANGED_CELLS = 45   # diff explosion => our anchor jumped
TRACK_EXPIRY_NEAR = 6         # unconfirmed ticks for a track within cheb 4
                              # (extended while in_combat: fight evidence)
TRACK_EXPIRY_FAR = 20         # unconfirmed ticks before a far track dies
TRACK_EXPIRY_OUT = 40         # ... once outside the window entirely


@dataclass
class Track:
    pos: tuple[int, int]          # frame coords
    kind: str                     # "melee" | "bow" | "player"
    delta: int                    # last seen delta bucket (enemies)
    hp_bucket: int
    element: int
    last_confirmed: int           # tick of last byte-change confirmation
    born: int
    inferred: bool = False        # position from model, not observation
    obs_pos: tuple[int, int] = (0, 0)   # last OBSERVED position


class WorldModel:
    def __init__(self):
        self.reset(0)

    def reset(self, tick: int) -> None:
        self.origin_tick = tick
        self.pos = (0, 0)                 # our frame position
        self.tracks: list[Track] = []
        self.prev_tiles: np.ndarray | None = None
        self.last_action: int = 4
        self.teleported = False

    # -- per-tick update -------------------------------------------------

    def observe(self, p: Percept, tick: int) -> None:
        self._in_combat = p.in_combat
        """Advance the model with this tick's percept. Call exactly once
        per tick, BEFORE reading any view, AFTER a life reset was
        handled by calling reset()."""
        shift = self._own_shift(p)
        self.teleported = False

        if self.prev_tiles is None:
            changed = np.zeros((WINDOW_ROWS, WINDOW_COLS), dtype=bool)
        else:
            changed = (p.tiles[:, :, 4:]
                       != self.prev_tiles[:, :, 4:]).any(axis=2)
            changed[CENTER_ROW, CENTER_COL] = False   # that's us
        n_changed = int(changed.sum())

        # teleportitis: the window content jumps wholesale; our anchor
        # and every track are garbage. Start clean.
        if n_changed >= TELEPORT_CHANGED_CELLS:
            self.prev_tiles = p.tiles.copy()
            self.tracks = []
            self.pos = (0, 0)
            self.origin_tick = tick
            self.teleported = True
            return

        self.pos = (self.pos[0] + shift[0], self.pos[1] + shift[1])

        # 1. model-propagate melee tracks that are inside our aggro
        # neighborhood: they chase deterministically - but only when WE
        # are unambiguously the chase target. Enemies attack the first
        # player in their row-major scan box; if another player track is
        # near the enemy, its behavior is unpredictable: hold position.
        # A predicted step onto impassable terrain stalls (move() fails).
        player_pos = [t.pos for t in self.tracks if t.kind == "player"]
        for t in self.tracks:
            t.inferred = True
            if t.kind == "melee":
                w = self._to_window(t.pos)
                if w is not None and self._cheb(w) <= NPC_AGGRO:
                    contested = any(self._dist(t.pos, pp) <= 5
                                    for pp in player_pos)
                    if contested:
                        continue
                    np_pos = self._chase_step(t.pos)
                    nw = self._to_window(np_pos)
                    if nw is None or p.passable(nw[0], nw[1]):
                        t.pos = np_pos

        # 2. observation update: every changed cell with entity bytes is
        # a confirmed presence NOW. Match to nearest track (<= 2 frame
        # cells - one own move + one enemy move), else new track.
        obs_hits = []
        for r, c in zip(*np.nonzero(changed)):
            ent_type = int(p.tiles[r, c, 4])
            if ent_type == 0:
                continue
            anim = int(p.tiles[r, c, 8])
            if anim == ANIM_DEATH:
                # corpse frame: kill the matching track if any
                fpos = self._to_frame((int(r), int(c)))
                self.tracks = [t for t in self.tracks
                               if self._dist(t.pos, fpos) > 1]
                continue
            obs_hits.append((int(r), int(c), ent_type))

        claimed: set[int] = set()
        for r, c, ent_type in obs_hits:
            fpos = self._to_frame((r, c))
            kind = ("player" if ent_type == ENTITY_PLAYER else
                    ("bow" if int(p.tiles[r, c, 5]) != 0 else "melee"))
            best_i, best_d = None, 99
            for i, t in enumerate(self.tracks):
                if i in claimed or t.kind != kind:
                    continue
                d = self._dist(t.pos, fpos)
                if d <= 2 and d < best_d:
                    best_i, best_d = i, d
            if best_i is not None:
                t = self.tracks[best_i]
                claimed.add(best_i)
                t.pos = fpos
                t.obs_pos = fpos
                t.inferred = False
                t.last_confirmed = tick
                if kind != "player":
                    t.delta = int(p.tiles[r, c, 6])
                    t.hp_bucket = int(p.tiles[r, c, 7])
                    t.element = int(p.tiles[r, c, 5])
            else:
                self.tracks.append(Track(
                    pos=fpos, obs_pos=fpos, kind=kind,
                    delta=int(p.tiles[r, c, 6]),
                    hp_bucket=int(p.tiles[r, c, 7]),
                    element=int(p.tiles[r, c, 5]),
                    last_confirmed=tick, born=tick, inferred=False))

        # 3. expiry + dedup. Near tracks are the dangerous ones AND the
        # phantom-prone ones: a chase-propagated track whose enemy left
        # converges on us and would sit forever. Evidence rules:
        #  - a track at Manhattan-1 while we are NOT in combat for 2+
        #    ticks is provably absent (adjacent melee attack every tick)
        #  - otherwise near tracks live TRACK_EXPIRY_NEAR unconfirmed
        #    ticks (extended while in_combat: the fight is the proof)
        kept: list[Track] = []
        for t in self.tracks:
            w = self._to_window(t.pos)
            age = tick - t.last_confirmed
            if w is None:
                if age <= TRACK_EXPIRY_OUT:
                    kept.append(t)
                continue
            man = abs(w[0] - CENTER_ROW) + abs(w[1] - CENTER_COL)
            near_player = any(self._dist(t.pos, pp) <= 2
                              for pp in player_pos)
            if t.kind == "melee" and man == 1 and not p.in_combat \
                    and age >= 2 and not near_player:
                continue
            # an inferred melee track at point-blank range that has not
            # been confirmed for several ticks is a phantom: a real one
            # would be attacking (in_combat) or moving (diff-visible)
            if t.kind == "melee" and self._cheb(w) <= 2 and t.inferred \
                    and age >= 3 and not p.in_combat and not near_player:
                continue
            # soft-confirm: a stationary enemy writes byte-identical
            # imprints (diff-invisible); if the cell still carries this
            # track's exact signature, extend its lease
            sig_ok = (int(p.tiles[w[0], w[1], 4]) != 0
                      and int(p.tiles[w[0], w[1], 5]) == t.element
                      and int(p.tiles[w[0], w[1], 7]) == t.hp_bucket)
            if self._cheb(w) <= 4:
                limit = 20 if p.in_combat else (
                    14 if sig_ok else TRACK_EXPIRY_NEAR)
            else:
                limit = TRACK_EXPIRY_FAR
            if age <= limit:
                kept.append(t)
        # dedup: two same-kind tracks on one cell merge (keep fresher)
        by_cell: dict[tuple[int, int, str], Track] = {}
        for t in kept:
            key = (t.pos[0], t.pos[1], t.kind)
            cur = by_cell.get(key)
            if cur is None or t.last_confirmed > cur.last_confirmed:
                by_cell[key] = t
        self.tracks = list(by_cell.values())

        self.prev_tiles = p.tiles.copy()

    def note_action(self, action: int) -> None:
        """Record the action we RETURNED this tick (used by next tick's
        own-shift estimation)."""
        self.last_action = action

    # -- views -----------------------------------------------------------

    def enemies(self, kind: str | None = None):
        """[(wr, wc, track)] for enemy tracks currently in-window."""
        out = []
        for t in self.tracks:
            if t.kind == "player":
                continue
            if kind is not None and t.kind != kind:
                continue
            w = self._to_window(t.pos)
            if w is not None:
                out.append((w[0], w[1], t))
        return out

    def occupancy(self) -> set[tuple[int, int]]:
        """Window cells our move() would bounce off (all tracks)."""
        occ = set()
        for t in self.tracks:
            w = self._to_window(t.pos)
            if w is not None:
                occ.add(w)
        return occ

    # -- helpers ---------------------------------------------------------

    @staticmethod
    def _cheb(w: tuple[int, int]) -> int:
        return max(abs(w[0] - CENTER_ROW), abs(w[1] - CENTER_COL))

    @staticmethod
    def _dist(a: tuple[int, int], b: tuple[int, int]) -> int:
        return max(abs(a[0] - b[0]), abs(a[1] - b[1]))

    def _to_window(self, fpos) -> tuple[int, int] | None:
        r = fpos[0] - self.pos[0] + CENTER_ROW
        c = fpos[1] - self.pos[1] + CENTER_COL
        if 0 <= r < WINDOW_ROWS and 0 <= c < WINDOW_COLS:
            return (r, c)
        return None

    def _to_frame(self, w) -> tuple[int, int]:
        return (self.pos[0] + w[0] - CENTER_ROW,
                self.pos[1] + w[1] - CENTER_COL)

    def _chase_step(self, epos) -> tuple[int, int]:
        """Exact chase rule, one enemy move toward us (frame coords).
        Terrain/entity blocking is unknown here; the planner re-checks."""
        dr = self.pos[0] - epos[0]
        dc = self.pos[1] - epos[1]
        if abs(dr) + abs(dc) <= 1:
            return epos                       # adjacent: it attacks
        if abs(dr) > abs(dc):
            return (epos[0] + (1 if dr > 0 else -1), epos[1])
        return (epos[0], epos[1] + (1 if dc > 0 else -1))

    def _own_shift(self, p: Percept) -> tuple[int, int]:
        walk = self.last_action
        if walk - RUN_OFFSET in MOVE_DELTAS:
            walk -= RUN_OFFSET
        if walk not in MOVE_DELTAS:
            return (0, 0)
        dr, dc = MOVE_DELTAS[walk]
        if p.anim == ANIM_MOVE:
            return (dr, dc)
        if p.anim == ANIM_RUN:
            return (2 * dr, 2 * dc)
        return (0, 0)

File: players/nimagent/world.nim
# L1 WorldModel: object tracks from the egocentric obs stream.
# Port of players/dancer/world.py with every session-4 verdict baked in
# (see cogames/nmmo learnings/dancer-v4-verdicts.md):
#
# - The obs entity bytes are WINDOW-anchored residue (a stale obs
#   buffer): never frame-align the diff, never trust imprints without a
#   track.
# - Tracks are created/confirmed ONLY by an observed byte change;
#   between confirmations melee tracks propagate by the exact chase
#   model (verified 33/33) and age out.
# - Absence proofs RE-ANCHOR to the last observed position instead of
#   deleting (chase propagation drifts on unseen blockers).
# - Ghost-strike inference: hp dropped with no adjacent melee track and
#   no aligned bow -> materialize an inferred melee from adjacent
#   residue (the byte-identical-rewrite blindness is real).

import std/tables
import percept

const
  NpcAggro* = 4
  TeleportChangedCells = 45
  TrackExpiryNear = 6
  TrackExpiryFar = 20
  TrackExpiryOut = 40

type
  TrackKind* = enum
    tkMelee = "melee", tkBow = "bow", tkPlayer = "player"

  Track* = ref object
    pos*: tuple[r, c: int]        # frame coords
    obsPos*: tuple[r, c: int]     # last OBSERVED position
    kind*: TrackKind
    delta*: int
    hpBucket*: int
    element*: int
    lastConfirmed*: int
    born*: int
    inferred*: bool
    ghost*: bool                  # born from ghost-strike inference:
                                  # expires fast, no residue-signature
                                  # lease (its own residue would renew
                                  # it forever)

  WorldModel* = object
    pos*: tuple[r, c: int]        # our frame position
    tracks*: seq[Track]
    prevTiles: array[WindowRows * WindowCols * TileBytes, uint8]
    havePrev: bool
    lastAction*: int
    teleported*: bool
    prevHp: int
    originTick*: int
    stillTicks: int               # consecutive ticks with no own move
    lastCorpse*: tuple[tick, wr, wc: int]  # most recent corpse diff
                                  # (window coords at sighting)
    dangerAnchors*: seq[tuple[r, c: int]]  # frame positions of bows
                                  # that actually HIT us this life -
                                  # leashed shooters, static all life

proc reset*(w: var WorldModel, tick: int) =
  w.originTick = tick
  w.pos = (0, 0)
  w.tracks = @[]
  w.havePrev = false
  w.lastAction = AtnNoop
  w.teleported = false
  w.prevHp = -1
  w.stillTicks = 0
  w.dangerAnchors = @[]
  w.lastCorpse = (-999, 0, 0)

proc initWorldModel*(): WorldModel =
  result.reset(0)

# -- helpers -----------------------------------------------------------

proc cheb(a, b: tuple[r, c: int]): int =
  max(abs(a.r - b.r), abs(a.c - b.c))

proc chebW(w: tuple[r, c: int]): int =
  max(abs(w.r - CenterRow), abs(w.c - CenterCol))

proc toWindow*(w: WorldModel, fpos: tuple[r, c: int]):
    tuple[ok: bool, r, c: int] =
  let r = fpos.r - w.pos.r + CenterRow
  let c = fpos.c - w.pos.c + CenterCol
  if r >= 0 and r < WindowRows and c >= 0 and c < WindowCols:
    (true, r, c)
  else:
    (false, 0, 0)

proc toFrame*(w: WorldModel, r, c: int): tuple[r, c: int] =
  (w.pos.r + r - CenterRow, w.pos.c + c - CenterCol)

proc chaseStep(w: WorldModel, epos: tuple[r, c: int]): tuple[r, c: int] =
  ## Exact chase rule, one enemy move toward us (frame coords).
  let dr = w.pos.r - epos.r
  let dc = w.pos.c - epos.c
  if abs(dr) + abs(dc) <= 1:
    return epos                    # adjacent: it attacks
  if abs(dr) > abs(dc):
    (epos.r + (if dr > 0: 1 else: -1), epos.c)
  else:
    (epos.r, epos.c + (if dc > 0: 1 else: -1))

proc ownShift(w: WorldModel, p: Percept): tuple[r, c: int] =
  var walk = w.lastAction
  if isRun(walk): walk -= RunOffset
  if not isMove(walk): return (0, 0)
  let d = MoveDeltas[walk]
  if p.anim == AnimMove: (d.dr, d.dc)
  elif p.anim == AnimRun: (2 * d.dr, 2 * d.dc)
  else: (0, 0)

# -- per-tick update ---------------------------------------------------

proc observe*(w: var WorldModel, p: Percept, tick: int) =
  ## Advance the model with this tick's percept. Call exactly once per
  ## tick, BEFORE reading views, AFTER handling a life reset.
  let shift = w.ownShift(p)
  w.teleported = false
  if shift == (0, 0): inc w.stillTicks
  else: w.stillTicks = 0

  # entity-byte window diff (bytes 4..9 per cell); center cell is us
  var changed: array[WindowRows * WindowCols, bool]
  var nChanged = 0
  if w.havePrev:
    for r in 0 ..< WindowRows:
      for c in 0 ..< WindowCols:
        if r == CenterRow and c == CenterCol: continue
        let base = (r * WindowCols + c) * TileBytes
        for b in TbEntType .. TbEntDir:
          if p.obs[base + b] != w.prevTiles[base + b]:
            changed[r * WindowCols + c] = true
            inc nChanged
            break

  # teleportitis: window content jumps wholesale; start clean
  if nChanged >= TeleportChangedCells:
    copyMem(addr w.prevTiles[0], unsafeAddr p.obs[0],
            WindowRows * WindowCols * TileBytes)
    w.havePrev = true
    w.tracks = @[]
    w.pos = (0, 0)
    w.originTick = tick
    w.teleported = true
    w.prevHp = p.hp
    return

  w.pos = (w.pos.r + shift.r, w.pos.c + shift.c)

  # 1. chase-propagate melee tracks near us, unless another player
  # track contests the chase target (then hold position)
  var playerPos: seq[tuple[r, c: int]]
  for t in w.tracks:
    if t.kind == tkPlayer: playerPos.add t.pos
  for t in w.tracks:
    t.inferred = true
    if t.kind == tkMelee:
      let win = w.toWindow(t.pos)
      if win.ok and chebW((win.r, win.c)) <= NpcAggro:
        var contested = false
        for pp in playerPos:
          if cheb(t.pos, pp) <= 5: contested = true; break
        if contested: continue
        let np = w.chaseStep(t.pos)
        let nw = w.toWindow(np)
        if not nw.ok or p.passable(nw.r, nw.c):
          t.pos = np

  # 2. observation update: every changed cell with entity bytes is a
  # confirmed presence NOW; match <=2 frame cells to a same-kind track
  type ObsHit = tuple[r, c, entType: int]
  var hits: seq[ObsHit]
  for r in 0 ..< WindowRows:
    for c in 0 ..< WindowCols:
      if not changed[r * WindowCols + c]: continue
      let et = p.tile(r, c, TbEntType)
      if et == 0: continue
      if p.tile(r, c, TbEntAnim) == AnimDeath:
        # corpse frame: kill the matching track if any. Loot drops at
        # kill sites and despawns in 20 ticks - note the sighting.
        let fpos = w.toFrame(r, c)
        var kept: seq[Track]
        for t in w.tracks:
          if cheb(t.pos, fpos) > 1: kept.add t
        w.tracks = kept
        w.lastCorpse = (tick, r, c)
        continue
      hits.add (r, c, et)

  var claimed: seq[bool] = newSeq[bool](w.tracks.len)
  for h in hits:
    let fpos = w.toFrame(h.r, h.c)
    let kind =
      if h.entType == EntityPlayer: tkPlayer
      elif p.tile(h.r, h.c, TbEntElement) != 0: tkBow
      else: tkMelee
    var bestI = -1
    var bestD = 99
    for i, t in w.tracks:
      if claimed[i] or t.kind != kind: continue
      let d = cheb(t.pos, fpos)
      if d <= 2 and d < bestD:
        bestI = i; bestD = d
    if bestI >= 0:
      let t = w.tracks[bestI]
      claimed[bestI] = true
      t.pos = fpos
      t.obsPos = fpos
      t.inferred = false
      t.ghost = false
      t.lastConfirmed = tick
      if kind != tkPlayer:
        t.delta = p.tile(h.r, h.c, TbEntDelta)
        t.hpBucket = p.tile(h.r, h.c, TbEntHp)
        t.element = p.tile(h.r, h.c, TbEntElement)
    else:
      w.tracks.add Track(
        pos: fpos, obsPos: fpos, kind: kind,
        delta: p.tile(h.r, h.c, TbEntDelta),
        hpBucket: p.tile(h.r, h.c, TbEntHp),
        element: p.tile(h.r, h.c, TbEntElement),
        lastConfirmed: tick, born: tick, inferred: false)
      # tracks grow mid-loop; keep the claim mask parallel (a track
      # created by an earlier hit IS a legal match target for later
      # hits this tick - python-tracker semantics)
      claimed.add false

  # 3. expiry + re-anchor + dedup
  var kept: seq[Track]
  for t in w.tracks:
    let win = w.toWindow(t.pos)
    let age = tick - t.lastConfirmed
    if not win.ok:
      if age <= TrackExpiryOut: kept.add t
      continue
    let man = abs(win.r - CenterRow) + abs(win.c - CenterCol)
    var nearPlayer = false
    for pp in playerPos:
      if cheb(t.pos, pp) <= 2: nearPlayer = true; break
    # absence proofs disprove the PROPAGATED position, not the enemy:
    # re-anchor to obsPos; delete only when the proof fires there too.
    # MOTION-GATED: "an adjacent melee would be attacking (in_combat)"
    # is only evidence while WE are stationary - a chaser legally
    # alternates moves with a moving player and never attacks, so the
    # proof fired on real chasers during walk-flees and teleported
    # their tracks away (measured kill chain, session 5 probe).
    let stationary = w.stillTicks >= 2
    if t.kind == tkMelee and man == 1 and not p.inCombat and
        age >= 2 and not nearPlayer and stationary:
      if t.pos != t.obsPos:
        t.pos = t.obsPos
        kept.add t
      continue
    if t.kind == tkMelee and chebW((win.r, win.c)) <= 2 and
        t.inferred and age >= 3 and not p.inCombat and not nearPlayer and
        stationary:
      if t.pos != t.obsPos:
        t.pos = t.obsPos
        kept.add t
      continue
    # soft-confirm: a stationary enemy is diff-invisible; if the cell
    # still carries this track's exact signature, extend the lease
    let sigOk = p.tile(win.r, win.c, TbEntType) != 0 and
      p.tile(win.r, win.c, TbEntElement) == t.element and
      p.tile(win.r, win.c, TbEntHp) == t.hpBucket
    var limit =
      if chebW((win.r, win.c)) <= 4:
        if p.inCombat: 20
        elif sigOk: 14
        else: TrackExpiryNear
      else:
        TrackExpiryFar
    if t.ghost:
      # ghost-born tracks sit on their own residue: the signature
      # lease would renew them forever. Byte-confirmation clears the
      # flag; otherwise they live 4 ticks.
      limit = 4
    if age <= limit: kept.add t
  # dedup: same-kind tracks on one cell merge (keep fresher)
  var byCell = initTable[(int, int, TrackKind), Track]()
  for t in kept:
    let key = (t.pos.r, t.pos.c, t.kind)
    if key notin byCell or t.lastConfirmed > byCell[key].lastConfirmed:
      byCell[key] = t
  w.tracks = @[]
  for t in byCell.values: w.tracks.add t

  # ghost-strike inference: hp dropped, yet no tracked melee adjacent
  # and no tracked bow aligned in range - the attacker is
  # diff-invisible; materialize an inferred melee from adjacent residue
  if w.prevHp >= 0 and p.hp < w.prevHp:
    var adjMelee, bowAligned = false
    for t in w.tracks:
      let win = w.toWindow(t.pos)
      if not win.ok: continue
      if t.kind == tkMelee and
          abs(win.r - CenterRow) + abs(win.c - CenterCol) <= 1:
        adjMelee = true
      if t.kind == tkBow and chebW((win.r, win.c)) <= 4 and
          (win.r == CenterRow or win.c == CenterCol):
        bowAligned = true
        if t.pos notin w.dangerAnchors:
          w.dangerAnchors.add t.pos
    if not adjMelee and not bowAligned:
      for d in MoveDeltas:
        let r = CenterRow + d.dr
        let c = CenterCol + d.dc
        if p.tile(r, c, TbEntType) != EntityEnemy or
            p.tile(r, c, TbEntElement) != 0:
          continue
        let fpos = w.toFrame(r, c)
        var exists = false
        for t in w.tracks:
          if t.kind == tkMelee and t.pos == fpos: exists = true; break
        if exists: continue
        w.tracks.add Track(
          pos: fpos, obsPos: fpos, kind: tkMelee,
          delta: p.tile(r, c, TbEntDelta),
          hpBucket: p.tile(r, c, TbEntHp), element: 0,
          lastConfirmed: tick, born: tick, inferred: true,
          ghost: true)
      # bow ghost-strike: a stationary aligned bow shoots from
      # stillness with zero diffs (never tracked). After an
      # unexplained hit, aligned bow-signature residue within its
      # 4-range IS the shooter - materialize it.
      for dr in -4 .. 4:
        for dc in -4 .. 4:
          if dr != 0 and dc != 0: continue
          if dr == 0 and dc == 0: continue
          let r = CenterRow + dr
          let c = CenterCol + dc
          if r < 0 or r >= WindowRows or c < 0 or c >= WindowCols:
            continue
          if p.tile(r, c, TbEntType) != EntityEnemy or
              p.tile(r, c, TbEntElement) == 0 or
              p.tile(r, c, TbEntAnim) == AnimDeath:
            continue
          let fpos = w.toFrame(r, c)
          var exists = false
          for t in w.tracks:
            if t.kind == tkBow and cheb(t.pos, fpos) <= 1:
              exists = true; break
          if exists: continue
          w.tracks.add Track(
            pos: fpos, obsPos: fpos, kind: tkBow,
            delta: p.tile(r, c, TbEntDelta),
            hpBucket: p.tile(r, c, TbEntHp),
            element: p.tile(r, c, TbEntElement),
            lastConfirmed: tick, born: tick, inferred: true,
            ghost: true)
          if fpos notin w.dangerAnchors:
            w.dangerAnchors.add fpos
  w.prevHp = p.hp

  copyMem(addr w.prevTiles[0], unsafeAddr p.obs[0],
          WindowRows * WindowCols * TileBytes)
  w.havePrev = true

proc noteAction*(w: var WorldModel, action: int) =
  w.lastAction = action

# -- views -------------------------------------------------------------

iterator enemies*(w: WorldModel, kind = tkMelee):
    tuple[r, c: int, t: Track] =
  ## In-window enemy tracks of one kind (window coords).
  for t in w.tracks:
    if t.kind != kind: continue
    let win = w.toWindow(t.pos)
    if win.ok:
      yield (win.r, win.c, t)

iterator enemiesExtended*(w: WorldModel, kind = tkBow, margin = 4):
    tuple[r, c: int, t: Track] =
  ## Like enemies() but includes tracks up to margin cells OUTSIDE the
  ## window (coords may exceed bounds) - run-safety for shallow rows.
  for t in w.tracks:
    if t.kind != kind: continue
    let r = t.pos.r - w.pos.r + CenterRow
    let c = t.pos.c - w.pos.c + CenterCol
    if r >= -margin and r < WindowRows + margin and
        c >= -margin and c < WindowCols + margin:
      yield (r, c, t)

proc occupancy*(w: WorldModel): seq[tuple[r, c: int]] =
  ## Window cells our move() would bounce off (all tracks).
  for t in w.tracks:
    let win = w.toWindow(t.pos)
    if win.ok: result.add (win.r, win.c)
